hcl2template/parser.go

```diff
@@ -295,10 +297,224 @@ func filterVarsFromLogs(inputOrLocal Variables) {
+			//
+			// Note: this is kinda brittle, we should understand scopes to accurately
+			// mark something from an expression as a reference to a local variable.
+			// No real good solution for this now, besides maybe forbidding something
+			// to be locally aliased as `local`.
+			if len(dep) < 2 {
+				continue
+			}
+			rs, err := NewRefStringFromDep(dep)
+			if err != nil {
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "failed to process local dependency",
+					Detail: fmt.Sprintf("An error occurred while processing a dependency for local variable %s: %s",
+						loc.LocalName, err),
+				})
+				continue
+			}
+
+			err = loc.RegisterDependency(rs)
+			if err != nil {
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "failed to register local dependency",
+					Detail: fmt.Sprintf("An error occurred while registering %q as a dependency for local variable %s: %s",
+						rs, loc.LocalName, err),
+				})
+			}
+		}
+	}
+
+	return diags
+}
+
+func (cfg *PackerConfig) buildPrereqsDAG() (*dag.AcyclicGraph, error) {
+	retGraph := dag.AcyclicGraph{}
+
+	verticesMap := map[string]dag.Vertex{}
+
+	// Do a first pass to create all the vertices
+	for ref := range cfg.Datasources {
+		// We keep a reference to the datasource separately from where it
+		// is used to avoid getting bit by the loop semantics.
+		//
+		// This `ds` local variable is the same object for every loop
+		// so if we directly use the address of this object, we'll end
+		// up referencing the last node of the loop for each vertex,
+		// leading to implicit cycles.
+		//
+		// However by capturing it locally in this loop, we have a
+		// reference to the actual datasource block, so it ends-up being
+		// the right instance for each vertex.
+		ds := cfg.Datasources[ref]
+		v := retGraph.Add(&ds)
+		verticesMap[fmt.Sprintf("data.%s", ds.Name())] = v
+	}
+	// Note: locals being references to the objects already, we can safely
+	// use the reference returned by the local loop.
+	for _, local := range cfg.LocalBlocks {
+		v := retGraph.Add(local)
+		verticesMap[fmt.Sprintf("local.%s", local.LocalName)] = v
+	}
+
+	// Connect the vertices together
+	//
+	// Vertices that don't have dependencies will be connected to the
+	// root vertex of the graph
+	for _, ds := range cfg.Datasources {
+		dsName := fmt.Sprintf("data.%s", ds.Name())
+
+		for _, dep := range ds.Dependencies {
+			retGraph.Connect(
+				dag.BasicEdge(verticesMap[dsName],
+					verticesMap[dep.String()]))
+		}
+	}
+	for _, loc := range cfg.LocalBlocks {
+		locName := fmt.Sprintf("local.%s", loc.LocalName)
+
+		for _, dep := range loc.dependencies {
+			retGraph.Connect(
+				dag.BasicEdge(verticesMap[locName],
+					verticesMap[dep.String()]))
+		}
+	}
+
+	if err := retGraph.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &retGraph, nil
+}
+
+func (cfg *PackerConfig) evaluateBuildPrereqs(skipDatasources bool) hcl.Diagnostics {
```

Comment:
The logic in this parser seems minimally unit tested, I notice that we have the packer_tests that invoke a larger section of the code, but I feel like it may be worth adding more testing to the logic in the parser, maybe we could make these functions public instead of private and invoke them in unit tests, just small tests that make sure we skip things aliased as local for example

Reply:
I agree it is not unit tested, and to be frank there are a bunch of pre-requisites to make this unit-testable, at the moment we rely on the `Initialize` tests (and subsequent phases) to test it, at least for regression checks.

Also to be clear: that code is in a `parser.go` file, but at this point we're not doing any kind of parsing, that was done before, it's kinda clumsy to have that logic in this file. I opted to add it here for this PR as this is where the phased code was already located, but honestly we should split this logic into separate files to avoid the confusion,